### DISCUSSION TITLE
Seeking Help to Enhance Huly Self-Hosting Capabilities

### DISCUSSION THREAD

**Discussion by aplatoff** [hcengineering/platform#4765] [General]
One of our goals is to refine the self-hosting capabilities of Huly. Currently, we support deployment through docker-compose, which our developers use daily for local testing. Additionally, we offer configurations for deploying Huly into a Kubernetes cluster. These two methods are our primary focus, but we're open to expanding our support based on valuable feedback.

Huly deployment comprises several components, including:

* Multiple stateless services (within docker containers)
* Integration-specific stateless services (e.g., for Google Mail, Google Calendar, Slack)
* MongoDB instance(s)
* Elasticsearch instance(s)
* Minio instance(s)

Deploying these components with docker-compose is straightforward, but we're eager to learn from both seasoned and novice self-hosters about best practices, especially for managing stateful services like MongoDB, Elasticsearch, and Minio.

Several questions revolve around Stateful Services:

1. **Shared Stateful Services**: Is there interest in a setup where multiple self-hosted products share instances of services like Minio or Elasticsearch to use resources more efficiently? Is this approach still relevant?
2. **Stateful Services Hosting**: Managing stateful services involves complexities related to persistent data, such as ensuring data availability after restarts and implementing backup strategies. How do you manage these challenges, and what best practices can you share? For example we manage stateful services in separate VMs out of Kubernetes cluster. I would not like to say this is a way to go, but is this a scenario some people would love to use as well?

Also, we're addressing the complexities of integrating with external services, for example Google Calendar, which is vital for Huly's Personal and Team planning features. Integrating with Google Calendar necessitates Google's permission to use their APIs—a process that is straightforward for our main deployment at [https://huly.io](https://huly.io). However, this integration might present challenges for self-hosted instances, including navigating legal and usage terms. We're interested in understanding how the self-hosting community approaches these scenarios.

Lastly, we're considering email notification services. Currently, we use AWS SES, but we understand the need for flexibility in self-hosted environments. Would support for multiple email services be beneficial (because sending email is quite complicated this days), or is an SMTP server sufficient?

Also what you guys think about supporting services like [elestio](https://elest.io) in addition to docker-compose and Kubernetes options for deployment?

We're committed to making Huly an excellent option for self-hosting and eagerly await your insights and suggestions.

Thank you! -- The Huly Team

**Comment by zeroone82** [comment]
Hi, I would like to answer your questions as someone who works as an SRE and also likes to self-host some projects.

Regarding stateful services:

1.  I think a lot of people would like this option, as they might already have deployed Minio or Elasticsearch and having central management would be useful. This could even be a requirement for enterprises looking to self-host.
I just want to be clear that in this case is not your Minio or Elasticsearch deployment that need to be enterprise ready, just give the option to use an external service that's already deployed.
2. Again, I would say volumes/backups should not be managed by the product/deployment itself but by the user/organization. I don't think there's inherently an advantage of running these services outside a kubernetes cluster if the nodes are correctly tainted and the deployments have affinity.


3. Regarding email, I think just being a client should be enough to send through most popular mail services including self-hosted smtp servers.
4. Supporting more options for deployment will never be a negative, as long as you don't forget to support and update the two main ways 95% of the users will want to deploy: Docker/docker-compose and Kubernetes (ideally through a Helm chart)

**Reply by siegenthalerroger** [reply]
I can only echo what @zeroone82 has said here. Huly should leave these "infrastructure" tasks to tools that do that job better. As such making sure that one can configure external URLs is the minimal requirement. For example  in our environment we already have all the dependencies available and problems you mentioned solved.

* MongoDB -> Mongo Community Operator and custom management tool for DB backup/restores
* Elasticsearch -> ELK Operator (or in our case we have Opensearch and its operator)
* Minio -> Technically we could have MinIO proxying S3 to Azure Blob Storage, however it'd be nice to not need that tbh
* Disks -> Velero does regular snapshots that can be application-consistent if configured

Concerning the Google Calendar/other API things. Having google cal (or mail) access actually is easy if you don't intend to have it outside of your own workspace. For most other services this is similar. For self-hosting I'd just leave this up to the user to make sure they have an API key, etc. that works. Nothing other services don't require from self-hosters.
